File: strings/165_compare_version_numbers.py

import unittest

"""
Given two version numbers, version1 and version2, compare them.

Version numbers consist of one or more revisions joined by a dot '.'. Each revision consists of digits and may contain leading zeros. Every revision contains at least one character. Revisions are 0-indexed from left to right, with the leftmost revision being revision 0, the next revision being revision 1, and so on. For example 2.5.33 and 0.1 are valid version numbers.

To compare version numbers, compare their revisions in left-to-right order. Revisions are compared using their integer value ignoring any leading zeros. This means that revisions 1 and 001 are considered equal. If a version number does not specify a revision at an index, then treat the revision as 0. For example, version 1.0 is less than version 1.1 because their revision 0s are the same, but their revision 1s are 0 and 1 respectively, and 0 < 1.

Return the following:

Constraints:

1 <= version1.length, version2.length <= 500
version1 and version2 only contain digits and '.'.
version1 and version2 are valid version numbers.
All the given revisions in version1 and version2 can be stored in a 32-bit integer.

"""

class Solution:
    def compareVersion(self, version1: str, version2: str) -> int:

        v1 = [int(num) for num in version1.split('.')]
        v2 = [int(num) for num in version2.split('.')]

        while len(v1) > len(v2):
            v2.append(0)
        while len(v2) > len(v1):
            v1.append(0)

        for i in range(len(v1)):
            if v1[i] > v2[i]:
                return 1
            if v1[i] < v2[i]:
                return -1

        return 0


class TestCompareVersion(unittest.TestCase):
    def test_compare_version_equal(self):
        sol = Solution()
        self.assertEqual(sol.compareVersion("1.0", "1.0"), 0)
        self.assertEqual(sol.compareVersion("1.2.3", "1.2.3"), 0)
        self.assertEqual(sol.compareVersion("0.0.1", "0.0.1"), 0)

    def test_compare_version_v1_greater(self):
        sol = Solution()
        self.assertEqual(sol.compareVersion("1.1", "1.0"), 1)
        self.assertEqual(sol.compareVersion("2.0", "1.1"), 1)
        self.assertEqual(sol.compareVersion("1.2.3", "1.2.2"), 1)
        self.assertEqual(sol.compareVersion("1.2.0", "1.1"), 1)

    def test_compare_version_v2_greater(self):
        sol = Solution()
        self.assertEqual(sol.compareVersion("1.0", "1.1"), -1)
        self.assertEqual(sol.compareVersion("1.1", "2.0"), -1)
        self.assertEqual(sol.compareVersion("1.2.2", "1.2.3"), -1)
        self.assertEqual(sol.compareVersion("1.1", "1.2.0"), -1)

    def test_compare_version_different_length(self):
        sol = Solution()
        self.assertEqual(sol.compareVersion("1", "1.0"), 0)
        self.assertEqual(sol.compareVersion("1", "1.1"), -1)
        self.assertEqual(sol.compareVersion("1.0", "1"), 0)
        self.assertEqual(sol.compareVersion("1.1", "1"), 1)


if __name__ == '__main__':
    unittest.main()

File: strings/165_compare_version_numbers.rs
impl Solution {
    pub fn compare_version(version1: String, version2: String) -> i32 {

        let v1: Vec<i32> = version1.split('.').map(|s| s.parse().unwrap_or(0)).collect();
        let v2: Vec<i32> = version2.split('.').map(|s| s.parse().unwrap_or(0)).collect();

        let max_len = v1.len().max(v2.len());

        for i in 0..max_len {
            let num1 = v1.get(i).unwrap_or(&0);
            let num2 = v2.get(i).unwrap_or(&0);

            if num1 > num2 {
                return 1;
            } else if num1 < num2 {
                return -1;
            }
        }

        0
    }

}

// Old solution
// impl Solution {
//     pub fn compare_version(version1: String, version2: String) -> i32 {
//
//         let mut v1: Vec<&str> = version1.split('.').collect();
//         let mut v2: Vec<&str> = version2.split('.').collect();
//         let mut v1i: Vec<i32> = Vec::new();
//         let mut v2i: Vec<i32> = Vec::new();
//
//         for version in v1.iter() {
//             let v = version.parse().unwrap();
//             v1i.push(v);
//         }
//
//         for version in v2.iter() {
//             let v = version.parse().unwrap();
//             v2i.push(v);
//         }
//
//         while v1i.len() < v2i.len() {
//             v1i.push(0);
//         }
//
//         while v2i.len() < v1i.len() {
//             v2i.push(0);
//         }
//
//         for i in 0..v1i.len() {
//             if v1i[i] > v2i[i] {
//                 return 1;
//             }
//             if v1i[i] < v2i[i] {
//                 return -1;
//             }
//         }
//
//         0
//     }
// }